File: python/src/tensor_op/example/attention_post.py
#!/usr/bin/env python3
# coding: utf-8
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This file is a part of the CANN Open Software.
# Licensed under CANN Open Software License Agreement Version 1.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# ======================================================================================================================
"""
"""
from dataclasses import dataclass
import pto


SHAPE_DIM4 = 4
SHAPE_DIM3 = 3
SHAPE_DIM2 = 2


@dataclass
class PostTileConfig:
    tile_b: int = 8
    tile_s: int = 1


@dataclass
class PostTensors:
    weight_u_v: pto.tensor = pto.tensor()
    weight_o: pto.tensor = pto.tensor()
    weight_uv_scale: pto.tensor = pto.tensor()
    smooth_scales_w_uv: pto.tensor = pto.tensor()
    weight_o_scale: pto.tensor = pto.tensor()
    smooth_scales_wo: pto.tensor = pto.tensor()


def post_compute(
        input_tensor: pto.tensor,
        post_tensors: PostTensors,
        tile_config: PostTileConfig,
        post_out: pto.tensor,
):

    if not (len(input_tensor.shape) == SHAPE_DIM4 and
        len(post_tensors.weight_u_v.shape) == SHAPE_DIM3 and
        len(post_tensors.weight_o.shape) == SHAPE_DIM2):
        raise ValueError("Tensor shapes do not match the expected dimensions")
    dtype = input_tensor.dtype
    n = post_tensors.weight_u_v.shape[0]
    kv_lora_rank = post_tensors.weight_u_v.shape[1]
    v_head_dim = post_tensors.weight_u_v.shape[2]
    h = post_tensors.weight_o.shape[1]

    tile_b = tile_config.tile_b
    tile_s = tile_config.tile_s
    tile_b_s = tile_b * tile_s

    is_quant_w_uv = (post_tensors.weight_uv_scale.has_storage()
                        if post_tensors.weight_uv_scale is not None else False)
    is_smooth_w_uv = (post_tensors.smooth_scales_w_uv.has_storage()
                        if post_tensors.smooth_scales_w_uv is not None else False)
    is_quant_wo = (post_tensors.weight_o_scale.has_storage()
                        if post_tensors.weight_o_scale is not None else False)
    is_smooth_wo = (post_tensors.smooth_scales_wo.has_storage()
                        if post_tensors.smooth_scales_wo is not None else False)

    b = input_tensor.shape[0];  # SymbolicScalar b = GetInputShapeDim(input_tensor, 0);
    s = input_tensor.shape[1];  # SymbolicScalar s = GetInputShapeDim(input_tensor, 1);
    b_loop = pto.symbolic_scalar(b // tile_b)
    s_loop = pto.symbolic_scalar(s // tile_s)

    for b_idx in pto.loop(0, b_loop, 1, name="POST_LOOP_L0_bIdx", idx_name="b_idx", submit_before_loop=True):
        def inside_b_idx_loop_attention_post(b_idx):
            nonlocal tile_s, tile_b, kv_lora_rank, tile_b_s, v_head_dim, n
            b_offset = b_idx * tile_b
            for s_idx in pto.loop(0, s_loop, 1, name="POST_LOOP_L1_sIdx", idx_name="s_idx"):
                def inside_s_idx_loop_attention_post(s_idx):
                    nonlocal tile_s, tile_b, kv_lora_rank, tile_b_s, v_head_dim, b_offset, n
                    s_offset = s_idx * tile_s
                    out_offset = [b_offset, s_offset, pto.symbolic_scalar(0)]

                    pto.set_vec_tile_shapes(*[1, 1, 32, kv_lora_rank])
                    input_view = pto.view(input_tensor, [tile_b, tile_s, n, kv_lora_rank], [b_offset, s_offset, 0, 0])
                    pto.set_semantic_label("postReshape1")
                    input_res = pto.reshape(input_view, [tile_b_s, n, kv_lora_rank])
                    pto.set_vec_tile_shapes(*[min(32, tile_b_s), 2, kv_lora_rank])
                    pto.set_semantic_label("postTranspose1")
                    input_trans = pto.transpose(input_res, [0, 1]);  # [n,tileBS,kvLoraRank]

                    pto.set_semantic_label("postBmm")
                    c0 = 16
                    m = (min(32, tile_b_s) + c0 - 1) // c0 * c0
                    pto.set_cube_tile_shapes([m, m], [min(256, kv_lora_rank),
                                        min(512, kv_lora_rank)], [v_head_dim, v_head_dim], True)

                    bmm = pto.tensor()
                    if is_quant_w_uv:
                        def inside_of_if_is_quant_w_uv():
                            nonlocal bmm, kv_lora_rank
                            pto.set_semantic_label("postQuantWUv")
                            pto.set_vec_tile_shapes(*[1, 1, min(512, kv_lora_rank)])
                            quant_res = None
                            if is_smooth_w_uv:
                                quant_res = pto.quant(input_trans, True, True, post_tensors.smooth_scales_w_uv)
                            else:
                                quant_res = pto.quant(input_trans, True, False)
                            input_trans_quant = quant_res[0]
                            scale_dequant = quant_res[1]

                            mm = pto.matmul(input_trans_quant, post_tensors.weight_u_v, pto.DT_INT32)

                            pto.set_semantic_label("postDequantWUv")
                            pto.set_vec_tile_shapes(*[1, min(16, tile_b_s), min(32, v_head_dim)])
                            res = pto.cast(mm, pto.DT_FP32)
                            res[:] = pto.mul(res, scale_dequant)
                            res[:] = pto.mul(res, post_tensors.weight_uv_scale)
                            bmm[:] = pto.cast(res, dtype, pto.cast_mode.CAST_RINT)
                        inside_of_if_is_quant_w_uv()
                    else:
                        bmm[:] = pto.matmul(input_trans, post_tensors.weight_u_v, dtype)

                    pto.set_semantic_label("postTranspose2")
                    pto.set_vec_tile_shapes(*[4, min(32, tile_b_s), v_head_dim])
                    bmm_trans = pto.transpose(bmm, [0, 1])
                    pto.set_semantic_label("postReshape2")
                    bmm_res = pto.reshape(bmm_trans, [tile_b_s, n * v_head_dim])

                    mm_res: pto.tensor = pto.tensor()
                    pto.set_cube_tile_shapes([m, m], [min(512, n * v_head_dim), min(512, n * v_head_dim)],
                                                [min(64, h), min(64, h)], True)
                    if is_quant_wo:
                        def inside_of_if_is_quant_wo():
                            pto.set_semantic_label("postQuantWo")
                            pto.set_vec_tile_shapes(*[1, n * v_head_dim])
                            quant_res = None
                            if is_smooth_wo:
                                quant_res = pto.quant(bmm_res, True, True, post_tensors.smooth_scales_wo)
                            else:
                                quant_res = pto.quant(bmm_res, True, False)
                            bmm_res_quant = quant_res[0]
                            scale_dequant = quant_res[1]

                            pto.set_semantic_label("postMm")
                            mm = pto.matmul(bmm_res_quant, post_tensors.weight_o, pto.DT_INT32)

                            pto.set_semantic_label("postDequantWo")
                            pto.set_vec_tile_shapes(*[min(32, tile_b_s), min(32, h)])
                            res = pto.cast(mm, pto.DT_FP32)
                            res[:] = pto.mul(res, scale_dequant);   # [tileBS, h] * [tileBS, 1] -> [tileBS, h]
                            res[:] = pto.mul(res, post_tensors.weight_o_scale)   # [tileBS, h] * [1, h] -> [tileBS, h]
                            mm_res[:] = pto.cast(res, dtype, pto.cast_mode.CAST_RINT)
                        inside_of_if_is_quant_wo()
                    else:
                        mm_res[:] = pto.matmul(bmm_res, post_tensors.weight_o, dtype)
                    pto.set_semantic_label("postReshape3")
                    post_out_view = pto.reshape(mm_res, [tile_b, tile_s, h])
                    pto.set_vec_tile_shapes(*[1, 1, h])
                    pto.assemble(post_out_view, out_offset, post_out)
                inside_s_idx_loop_attention_post(s_idx)
        inside_b_idx_loop_attention_post(b_idx)


def attention_post_standalone(
        input_tensor: pto.tensor,
        post_tensors: PostTensors,
        tile_config: PostTileConfig,
        post_out: pto.tensor,
):

    input_tensors = [input_tensor, post_tensors.weight_u_v, post_tensors.weight_o,
                        post_tensors.weight_uv_scale, post_tensors.smooth_scales_w_uv,
                            post_tensors.weight_o_scale, post_tensors.smooth_scales_wo]
    output_tensors = [post_out]
    with pto.function("POST_MAIN", input_tensors, output_tensors):
        post_compute(input_tensor, post_tensors, tile_config, post_out)

## UT


@dataclass
class TestPostParams:
    b: int
    n: int
    s: int
    h: int
    kv_lora_rank: int
    v_head_dim: int


def test_attention_post_ut(**kwargs):
    params = kwargs.get("params")
    tile_config = kwargs.get("tile_config")
    d_type = kwargs.get("d_type", pto.DT_FP16)
    nz = kwargs.get("nz", True)
    w_uv_dtype = kwargs.get("w_uv_dtype", pto.DT_INT8)
    is_smooth_wuv = kwargs.get("is_smooth_wuv", False)
    w_o_dtype = kwargs.get("w_o_dtype", pto.DT_INT8)
    is_smooth_wo = kwargs.get("is_smooth_wo", False)

    b = params.b
    n = params.n
    s = params.s
    h = params.h
    kv_lora_rank = params.kv_lora_rank
    v_head_dim = params.v_head_dim

    d_type = pto.DT_FP16 if d_type == pto.DT_FP16 else pto.DT_BF16
    is_quant_w_uv = w_uv_dtype == pto.DT_INT8
    is_quant_wo = w_o_dtype == pto.DT_INT8

    x_shape = [b, s, n, kv_lora_rank]
    w_uv_shape = [n, kv_lora_rank, v_head_dim]
    w_uv_scale_shape = [n, 1, v_head_dim]
    smooth_w_uv_shape = [1, kv_lora_rank]
    wo_shape = [n * v_head_dim, h]
    wo_scale_shape = [1, h]
    smooth_wo_shape = [1, n * v_head_dim]
    out_shape = [b, s, h]

    weight_format = pto.TileOpFormat.TILEOP_NZ if nz else pto.TileOpFormat.TILEOP_ND
    x = pto.Tensor(x_shape, d_type, "x")
    w_uv = pto.Tensor(w_uv_shape, pto.DT_INT8 if is_quant_w_uv else d_type, "wUv")
    w_uv_scale = pto.Tensor()
    smooth_w_uv = pto.Tensor()
    wo = pto.Tensor(wo_shape, pto.DT_INT8 if is_quant_wo else d_type, "wo", weight_format)
    wo_scale = pto.Tensor()
    smooth_wo = pto.Tensor()
    post_out = pto.Tensor(out_shape, d_type, "postOut")

    if is_quant_w_uv:
        scale = pto.Tensor(w_uv_scale_shape, pto.DT_FP32, "wUvScale")
        w_uv_scale[:] = scale
        if is_smooth_wuv:
            smooth = pto.Tensor(smooth_w_uv_shape, pto.DT_FP32, "smoothWUv")
            smooth_w_uv[:] = smooth

    if is_quant_wo:
        scale = pto.Tensor(wo_scale_shape, pto.DT_FP32, "woScale")
        wo_scale[:] = scale
        if is_smooth_wo:
            smooth = pto.Tensor(smooth_wo_shape, pto.DT_FP32, "smoothWo")
            smooth_wo[:] = smooth

    post_tensors = PostTensors(w_uv, wo, w_uv_scale, smooth_w_uv, wo_scale, smooth_wo)
    attention_post_standalone(x, post_tensors, tile_config, post_out)


if __name__ == "__main__":
    params = TestPostParams(32, 128, 2, 7168, 512, 128)
    tile_config = PostTileConfig(16, 1)
    d_type = pto.DT_BF16
    nz = True
    w_uv_dtype = pto.DT_FP16
    is_smooth_wuv = False
    w_o_dtype = pto.DT_INT8
    is_smooth_wo = True

    test_attention_post_ut(params=params, tile_config=tile_config, nz=True, d_type=d_type, is_smooth_wuv=is_smooth_wuv,
                            w_uv_dtype=w_uv_dtype, w_o_dtype=w_o_dtype, is_smooth_wo=is_smooth_wo)

File: python/example/cpp_reference/src/attention_post.cpp
/**
 * Copyright (c) 2025 Huawei Technologies Co., Ltd.
 * This file is a part of the CANN Open Software.
 * Licensed under CANN Open Software License Agreement Version 1.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

/*!
 * \file attention_post.cpp
 * \brief
 */
#include "interface/function/function.h"
#include "interface/configs/config_manager.h"
using namespace npu::tile_fwk;

struct PostTileConfig {
    int tileB = 8;  // tileB is 8
    int tileS = 1;
};

struct PostTensors {
    Tensor weightUV;
    Tensor weightO;
    Tensor weightUvScale;
    Tensor smoothScalesWUv;
    Tensor weightOScale;
    Tensor smoothScalesWo;
};

namespace npu::tile_fwk {

// b and s is dynamic, support:
// b: 16, 32, 64, 24, 48, 96
// s: 1, 2
void PostCompute(Tensor &input, PostTensors &postTensors, const PostTileConfig &tileConfig, Tensor &postOut) {
    // input: [b,s,n,kvLoraRank], fp16/bf16
    // weightUV: [n,kvLoraRank,vHeadDim], fp16/bf16/int8
    // weightUvScale: [n,1,kvLoraRank], fp32
    // weightO: [v*kvLoraRank,h], fp16/bf16/int8
    // weightOScale: [1,h], fp32
    // params check
    assert(input->shape.size() == SHAPE_DIM4 && postTensors.weightUV->shape.size() == SHAPE_DIM3 &&
            postTensors.weightO->shape.size() == SHAPE_DIM2);
    auto dtype = input->Datatype();
    auto n = postTensors.weightUV->shape[0];
    auto kvLoraRank = postTensors.weightUV->shape[1];
    auto vHeadDim = postTensors.weightUV->shape[2];
    auto h = postTensors.weightO->shape[1];

    int tileB = tileConfig.tileB;
    int tileS = tileConfig.tileS;
    int tileBS = tileB * tileS;

    bool isQuantWUv = postTensors.weightUvScale.GetStorage() != nullptr;
    bool isSmoothWUv = postTensors.smoothScalesWUv.GetStorage() != nullptr;
    bool isQuantWo = postTensors.weightOScale.GetStorage() != nullptr;
    bool isSmoothWo = postTensors.smoothScalesWo.GetStorage() != nullptr;

    int b = input->shape[0];  // SymbolicScalar b = GetInputShapeDim(input, 0);
    int s = input->shape[1];  // SymbolicScalar s = GetInputShapeDim(input, 1);
    SymbolicScalar bLoop = b / tileB;
    SymbolicScalar sLoop = s / tileS;

    LOOP("POST_LOOP_L0_bIdx", FunctionType::DYNAMIC_LOOP, bIdx, LoopRange(0, bLoop, 1), {}, true) {
        SymbolicScalar bOffset = bIdx * tileB;
        LOOP("POST_LOOP_L1_sIdx", FunctionType::DYNAMIC_LOOP, sIdx, LoopRange(0, sLoop, 1)) {
            SymbolicScalar sOffset = sIdx * tileS;
            std::vector<SymbolicScalar> outOffset = {bOffset, sOffset, 0};

            TileShape::Current().SetVecTile({1, 1, 32, kvLoraRank});
            auto inputView = View(input, {tileB, tileS, n, kvLoraRank}, {bOffset, sOffset, 0, 0});
            config::SetSemanticLabel("postReshape1");
            auto inputRes = Reshape(inputView, {tileBS, n, kvLoraRank});
            TileShape::Current().SetVecTile({std::min(32, tileBS), 2, kvLoraRank});
            config::SetSemanticLabel("postTranspose1");
            auto inputTrans = Transpose(inputRes, {0, 1});  // [n,tileBS,kvLoraRank]

            config::SetSemanticLabel("postBmm");
            int c0 = 16;
            int m = (std::min(32, tileBS) + c0 - 1) / c0 * c0;
            TileShape::Current().SetCubeTile(
                {m, m}, {std::min(256L, kvLoraRank), std::min(512L, kvLoraRank)}, {vHeadDim, vHeadDim}, true);

            Tensor bmm;
            if (isQuantWUv) {
                config::SetSemanticLabel("postQuantWUv");
                TileShape::Current().SetVecTile({1, 1, std::min(512L, kvLoraRank)});
                std::tuple<Tensor, Tensor> quantRes;
                if(isSmoothWUv){
                    quantRes = Quant(inputTrans, true, true,  postTensors.smoothScalesWUv);
                }else{
                    quantRes = Quant(inputTrans, true, false);
                }
                auto inputTransQuant = std::get<0>(quantRes); // [n, tileBS, kvLoraRank], int8
                auto scaleDequant = std::get<1>(quantRes);    // [n, tileBS, 1], fp32

                // [n, tileBS, kvLoraRank] @ [n, kvLoraRank, vHeadDim] -> [n, tileBS, vHeadDim] int8 @ int8 -> int32
                auto mm = Matrix::BatchMatmul(DT_INT32, inputTransQuant,  postTensors.weightUV);

                config::SetSemanticLabel("postDequantWUv");
                TileShape::Current().SetVecTile({1, std::min(16, tileBS), std::min(32L, vHeadDim)});
                Tensor res = Cast(mm, DataType::DT_FP32);
                res = Mul(res, scaleDequant);  // [n, tileBS, VHeadDim] * [n, tileBS, 1] -> [n, tileBS, vHeadDim]
                res = Mul(res,  postTensors.weightUvScale); // [n, tileBS, VHeadDim] * [n, 1, VHeadDim] -> [n, tileBS, vHeadDim]
                bmm = Cast(res, dtype, CAST_RINT);
            } else {
                // [n,tileBS,kvLoraRank] @ [n,kvLoraRank,vHeadDim] -> [n,tileBS,vHeadDim]
                bmm = Matrix::BatchMatmul(dtype, inputTrans, postTensors.weightUV);
            }

            config::SetSemanticLabel("postTranspose2");
            TileShape::Current().SetVecTile({4, std::min(32, tileBS), vHeadDim});
            auto bmmTrans = Transpose(bmm, {0, 1}); // [n,tileBS,vHeadDim] -> [tileBS,n,vHeadDim]
            config::SetSemanticLabel("postReshape2");
            auto bmmRes = Reshape(bmmTrans, {tileBS, n * vHeadDim});

            Tensor mmRes;
            TileShape::Current().SetCubeTile({m, m}, {std::min(512L, n * vHeadDim), std::min(512L, n * vHeadDim)},
                {std::min(64L, h), std::min(64L, h)}, true);
            if (isQuantWo) {
                config::SetSemanticLabel("postQuantWo");
                TileShape::Current().SetVecTile({1, n * vHeadDim});
                std::tuple<Tensor, Tensor> quantRes;
                if (isSmoothWo) {
                    quantRes = Quant(bmmRes, true, true,  postTensors.smoothScalesWo);
                } else {
                    quantRes = Quant(bmmRes, true, false);
                }
                auto bmmResQuant = std::get<0>(quantRes);  // [tileBS, n*vHeadDim], int8
                auto scaleDequant = std::get<1>(quantRes); // [tileBS, 1], fp32

                config::SetSemanticLabel("postMm");
                // [tileBS, n*vHeadDim] @ [n*vHeadDim, h] -> [tileBS, h], int8 @ int8 -> int32
                Tensor mm = Matrix::Matmul(DataType::DT_INT32, bmmResQuant,  postTensors.weightO);

                config::SetSemanticLabel("postDequantWo");
                TileShape::Current().SetVecTile({std::min(32, tileBS), std::min(32L, h)});
                Tensor res = Cast(mm, DataType::DT_FP32);
                res = Mul(res, scaleDequant);   // [tileBS, h] * [tileBS, 1] -> [tileBS, h]
                res = Mul(res,  postTensors.weightOScale);   // [tileBS, h] * [1, h] -> [tileBS, h]
                mmRes = Cast(res, dtype, CAST_RINT);
            } else {
                // [tileBS, n*vHeadDim] @ [n*vHeadDim, h] -> [tileBS, h], dtype @ dtype -> dtype
                mmRes = Matrix::Matmul(dtype, bmmRes,  postTensors.weightO);
            }
            config::SetSemanticLabel("postReshape3");
            auto postOutView = Reshape(mmRes, {tileB, tileS, h});
            TileShape::Current().SetVecTile({1, 1, h});
            Assemble(postOutView, outOffset, postOut);
        }
    }
}

void AttentionPostStandalone(Tensor &input, PostTensors &postTensors, const PostTileConfig &tileConfig, Tensor &postOut) {
    FUNCTION("POST_MAIN",
        {input, postTensors.weightUV, postTensors.weightO, postTensors.weightUvScale, postTensors.smoothScalesWUv,
            postTensors.weightOScale, postTensors.smoothScalesWo}, {postOut}) {
        PostCompute(input, postTensors, tileConfig, postOut);
    }
}

} // namespace npu::tile_fwk

// UT
struct TestPostParams {
    int b;
    int n;
    int s;
    int h;
    int kvLoraRank;
    int vHeadDim;
};

template <typename T = npu::tile_fwk::float16, bool nz = true, typename wUvDtype = int8_t, bool isSmoothWuv = false,
    typename wODtype = int8_t, bool isSmoothWo = false>
void TestAttentionPostUt(const TestPostParams &params, const PostTileConfig &tileConfig) {
    int b = params.b;
    int n = params.n;
    int s = params.s;
    int h = params.h;
    int kvLoraRank = params.kvLoraRank;
    int vHeadDim = params.vHeadDim;

    DataType dType = (std::is_same<T, npu::tile_fwk::float16>::value) ? DT_FP16 : DT_BF16;
    bool isQuantWUv = std::is_same<wUvDtype, int8_t>::value;
    bool isQuantWo = std::is_same<wODtype, int8_t>::value;

    std::vector<int64_t> xShape = {b, s, n, kvLoraRank};
    std::vector<int64_t> wUvShape = {n, kvLoraRank, vHeadDim};
    std::vector<int64_t> wUvScaleShape = {n, 1, vHeadDim};
    std::vector<int64_t> smoothWUvShape = {1, kvLoraRank};
    std::vector<int64_t> woShape = {n * vHeadDim, h};
    std::vector<int64_t> woScaleShape = {1, h};
    std::vector<int64_t> smoothWoShape = {1, n * vHeadDim};
    std::vector<int64_t> outShape = {b, s, h};

    TileOpFormat weightFormat = nz ? TileOpFormat::TILEOP_NZ : TileOpFormat::TILEOP_ND;
    Tensor x(dType, xShape, "x");
    Tensor wUv(isQuantWUv ? DT_INT8 : dType, wUvShape, "wUv");
    Tensor wUvScale;
    Tensor smoothWUv;
    Tensor wo(isQuantWo ? DT_INT8 : dType, woShape, "wo", weightFormat);
    Tensor woScale;
    Tensor smoothWo;
    Tensor postOut(dType, outShape, "postOut");

    if(isQuantWUv) {
        Tensor scale(DT_FP32, wUvScaleShape, "wUvScale");
        wUvScale =scale;
        if(isSmoothWuv) {
            Tensor smooth(DT_FP32, smoothWUvShape, "smoothWUv");
            smoothWUv =smooth;
        }
    }
    if (isQuantWo) {
        Tensor scale(DT_FP32, woScaleShape, "woScale");
        woScale = scale;
        if (isSmoothWo) {
            Tensor smooth(DT_FP32, smoothWoShape, "smoothWo");
            smoothWo = smooth;
        }
    }

    PostTensors postTensors{wUv, wo, wUvScale, smoothWUv, woScale, smoothWo};
    AttentionPostStandalone(x, postTensors, tileConfig, postOut);
}

int main() {
    // b, n, s, h, kvLoraRank, vHeadDim
    TestPostParams params = {32, 128, 2, 7168, 512, 128};
    PostTileConfig tileConfig = {16, 1};

    TestAttentionPostUt<npu::tile_fwk::bfloat16, true, npu::tile_fwk::float16, false, int8_t, true>(params, tileConfig);
    return 0;
}